# ENEO

[![Snakemake](https://img.shields.io/badge/snakemake-≤8.0.0-brightgreen.svg)](https://snakemake.github.io)
[![Linting](https://github.com/ctglab/ENEO/actions/workflows/formatting.yml/badge.svg?branch=main)](https://github.com/ctglab/ENEO/actions/workflows/formatting.yml)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-%23FE5196?logo=conventionalcommits&logoColor=white)](https://conventionalcommits.org)


ENEO (pronunced `/eˈnɛjo/`) is a Snakemake workflow developed for detecting immunogenic neoantigens arising from somatic mutations using only bulk tumor RNA-seq, without requiring matching samples or additional genomic data (WES/WGS). It uses a probabilistic model that leverages population genetics databases and genotype likelihoods to discriminate germline variants from the call set. Additional details are reported in the [bioRxiv preprint](https://www.biorxiv.org/content/10.1101/2024.08.08.607127v1).

## Usage

Pipeline documentation is available at [https://ctglab.github.io/ENEO](https://ctglab.github.io/ENEO)

## Cite

If you used this workflow in your work, don't forget to give credit to the authors by citing the original publication

>Tatoni, D., Dalsass, M., Brunelli, G., Grandi, G., Chiariello, M., & D'Aurizio, R. (2024). Efficient and effective identification of cancer neoantigens from tumor only RNA-seq. bioRxiv, 2024-08.






