
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://ctglab.github.io/ENEO/intervals/">
      
      
        <link rel="prev" href="../hpc/">
      
      
        <link rel="next" href="../resources/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Calling intervals - ENEO</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#calling-intervals" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="ENEO" class="md-header__button md-logo" aria-label="ENEO" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            ENEO
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Calling intervals
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/ctglab/ENEO" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    ENEO
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Introduction

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../hpc/" class="md-tabs__link">
        
  
  
    
  
  Run on HPC

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
  
    
  
  Calling intervals

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../resources/" class="md-tabs__link">
        
  
  
    
  
  Setup resources

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../setup/" class="md-tabs__link">
        
  
  
    
  
  Workflow Setup

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="ENEO" class="md-nav__button md-logo" aria-label="ENEO" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    ENEO
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/ctglab/ENEO" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    ENEO
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../hpc/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Run on HPC
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Calling intervals
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Calling intervals
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#generate-a-custom-interval-set" class="md-nav__link">
    <span class="md-ellipsis">
      Generate a custom interval set
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../resources/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Setup resources
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../setup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow Setup
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="calling-intervals">Calling intervals</h1>
<p>By default, ENEO performs variant calling on exons of protein coding genes, with the exception of two types of calling regions:</p>
<ul>
<li>
<p>Known hard-to-call regions of the human genome. These regions were <a href="https://www.biorxiv.org/content/10.1101/2023.10.27.563846v1.full">extensively profiled</a> by the GIAB consortium, and are <a href="https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/genome-stratifications/v3.5/GRCh38@all/">publicly available</a>. This regions are removed from the VCF file and so are not present in the final VCF output.</p>
</li>
<li>
<p>Genes involved in the antigen presentation: given the initial goal of the pipeline to be applied in a personalized cancer vaccine setup, we discared these genes as a source of unwanted variations. We obtained the set of genes using the KEGG pathway annotated as </p>
</li>
</ul>
<details class="example">
<summary>Expand to see the full set of excluded genes in ENEO</summary>
<table>
<thead>
<tr>
<th>symbol</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td>IFNG</td>
<td>interferon gamma</td>
</tr>
<tr>
<td>TNF</td>
<td>tumor necrosis factor</td>
</tr>
<tr>
<td>PSME1</td>
<td>proteasome activator subunit 1</td>
</tr>
<tr>
<td>PSME2</td>
<td>proteasome activator subunit 2</td>
</tr>
<tr>
<td>PSME3</td>
<td>proteasome activator subunit 3</td>
</tr>
<tr>
<td>HSPA8</td>
<td>heat shock protein family A (Hsp70) member 8</td>
</tr>
<tr>
<td>HSPA1A</td>
<td>heat shock protein family A (Hsp70) member 1A</td>
</tr>
<tr>
<td>HSPA1L</td>
<td>heat shock protein family A (Hsp70) member 1 like</td>
</tr>
<tr>
<td>HSPA1B</td>
<td>heat shock protein family A (Hsp70) member 1B</td>
</tr>
<tr>
<td>HSPA6</td>
<td>heat shock protein family A (Hsp70) member 6</td>
</tr>
<tr>
<td>HSPA2</td>
<td>heat shock protein family A (Hsp70) member 2</td>
</tr>
<tr>
<td>HSPA4</td>
<td>heat shock protein family A (Hsp70) member 4</td>
</tr>
<tr>
<td>HSP90AA1</td>
<td>heat shock protein 90 alpha family class A member 1</td>
</tr>
<tr>
<td>HSP90AB1</td>
<td>heat shock protein 90 alpha family class B member 1</td>
</tr>
<tr>
<td>HLA-A</td>
<td>major histocompatibility complex, class I, A</td>
</tr>
<tr>
<td>HLA-B</td>
<td>major histocompatibility complex, class I, B</td>
</tr>
<tr>
<td>HLA-C</td>
<td>major histocompatibility complex, class I, C</td>
</tr>
<tr>
<td>HLA-F</td>
<td>major histocompatibility complex, class I, F</td>
</tr>
<tr>
<td>HLA-G</td>
<td>major histocompatibility complex, class I, G</td>
</tr>
<tr>
<td>HLA-E</td>
<td>major histocompatibility complex, class I, E</td>
</tr>
<tr>
<td>HSPA5</td>
<td>heat shock protein family A (Hsp70) member 5</td>
</tr>
<tr>
<td>CANX</td>
<td>calnexin</td>
</tr>
<tr>
<td>B2M</td>
<td>beta-2-microglobulin</td>
</tr>
<tr>
<td>PDIA3</td>
<td>protein disulfide isomerase family A member 3</td>
</tr>
<tr>
<td>CALR</td>
<td>calreticulin</td>
</tr>
<tr>
<td>TAPBP</td>
<td>TAP binding protein</td>
</tr>
<tr>
<td>TAP1</td>
<td>transporter 1, ATP binding cassette subfamily B member</td>
</tr>
<tr>
<td>TAP2</td>
<td>transporter 2, ATP binding cassette subfamily B member</td>
</tr>
<tr>
<td>CD8A</td>
<td>CD8 subunit alpha</td>
</tr>
<tr>
<td>CD8B</td>
<td>CD8 subunit beta</td>
</tr>
<tr>
<td>CD8B2</td>
<td>CD8B family member 2</td>
</tr>
<tr>
<td>KIR3DL2</td>
<td>killer cell immunoglobulin like receptor, three Ig domains and long cytoplasmic tail 2</td>
</tr>
<tr>
<td>KIR3DL1</td>
<td>killer cell immunoglobulin like receptor, three Ig domains and long cytoplasmic tail 1</td>
</tr>
<tr>
<td>KIR3DL3</td>
<td>killer cell immunoglobulin like receptor, three Ig domains and long cytoplasmic tail 3</td>
</tr>
<tr>
<td>KIR3DS1</td>
<td>killer cell immunoglobulin like receptor, three Ig domains and short cytoplasmic tail 1</td>
</tr>
<tr>
<td>KIR2DL2</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and long cytoplasmic tail 2</td>
</tr>
<tr>
<td>KIR2DL1</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and long cytoplasmic tail 1</td>
</tr>
<tr>
<td>KIR2DL3</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and long cytoplasmic tail 3</td>
</tr>
<tr>
<td>KIR2DL4</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and long cytoplasmic tail 4</td>
</tr>
<tr>
<td>KIR2DL5A</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and long cytoplasmic tail 5A</td>
</tr>
<tr>
<td>KLRC1</td>
<td>killer cell lectin like receptor C1</td>
</tr>
<tr>
<td>KLRC2</td>
<td>killer cell lectin like receptor C2</td>
</tr>
<tr>
<td>KLRC3</td>
<td>killer cell lectin like receptor C3</td>
</tr>
<tr>
<td>KLRC4</td>
<td>killer cell lectin like receptor C4</td>
</tr>
<tr>
<td>KLRD1</td>
<td>killer cell lectin like receptor D1</td>
</tr>
<tr>
<td>KIR2DS1</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and short cytoplasmic tail 1</td>
</tr>
<tr>
<td>KIR2DS3</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and short cytoplasmic tail 3</td>
</tr>
<tr>
<td>KIR2DS4</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and short cytoplasmic tail 4</td>
</tr>
<tr>
<td>KIR2DS5</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and short cytoplasmic tail 5</td>
</tr>
<tr>
<td>KIR2DS2</td>
<td>killer cell immunoglobulin like receptor, two Ig domains and short cytoplasmic tail 2</td>
</tr>
<tr>
<td>IFI30</td>
<td>IFI30 lysosomal thiol reductase</td>
</tr>
<tr>
<td>LGMN</td>
<td>legumain</td>
</tr>
<tr>
<td>CTSB</td>
<td>cathepsin B</td>
</tr>
<tr>
<td>HLA-DMA</td>
<td>major histocompatibility complex, class II, DM alpha</td>
</tr>
<tr>
<td>HLA-DMB</td>
<td>major histocompatibility complex, class II, DM beta</td>
</tr>
<tr>
<td>HLA-DOA</td>
<td>major histocompatibility complex, class II, DO alpha</td>
</tr>
<tr>
<td>HLA-DOB</td>
<td>major histocompatibility complex, class II, DO beta</td>
</tr>
<tr>
<td>HLA-DPA1</td>
<td>major histocompatibility complex, class II, DP alpha 1</td>
</tr>
<tr>
<td>HLA-DPB1</td>
<td>major histocompatibility complex, class II, DP beta 1</td>
</tr>
<tr>
<td>HLA-DQA1</td>
<td>major histocompatibility complex, class II, DQ alpha 1</td>
</tr>
<tr>
<td>HLA-DQA2</td>
<td>major histocompatibility complex, class II, DQ alpha 2</td>
</tr>
<tr>
<td>HLA-DQB1</td>
<td>major histocompatibility complex, class II, DQ beta 1</td>
</tr>
<tr>
<td>HLA-DRA</td>
<td>major histocompatibility complex, class II, DR alpha</td>
</tr>
<tr>
<td>HLA-DRB1</td>
<td>major histocompatibility complex, class II, DR beta 1</td>
</tr>
<tr>
<td>HLA-DRB3</td>
<td>major histocompatibility complex, class II, DR beta 3</td>
</tr>
<tr>
<td>HLA-DRB4</td>
<td>major histocompatibility complex, class II, DR beta 4</td>
</tr>
<tr>
<td>HLA-DRB5</td>
<td>major histocompatibility complex, class II, DR beta 5</td>
</tr>
<tr>
<td>CD74</td>
<td>CD74 molecule</td>
</tr>
<tr>
<td>CTSL</td>
<td>cathepsin L</td>
</tr>
<tr>
<td>CTSS</td>
<td>cathepsin S</td>
</tr>
<tr>
<td>CD4</td>
<td>CD4 molecule</td>
</tr>
<tr>
<td>CIITA</td>
<td>class II major histocompatibility complex transactivator</td>
</tr>
<tr>
<td>RFX5</td>
<td>regulatory factor X5</td>
</tr>
<tr>
<td>RFXANK</td>
<td>regulatory factor X associated ankyrin containing protein</td>
</tr>
<tr>
<td>RFXAP</td>
<td>regulatory factor X associated protein</td>
</tr>
<tr>
<td>CREB1</td>
<td>cAMP responsive element binding protein 1</td>
</tr>
<tr>
<td>NFYA</td>
<td>nuclear transcription factor Y subunit alpha</td>
</tr>
<tr>
<td>NFYB</td>
<td>nuclear transcription factor Y subunit beta</td>
</tr>
<tr>
<td>NFYC</td>
<td>nuclear transcription factor Y subunit gamma</td>
</tr>
</tbody>
</table>
</details>
<h2 id="generate-a-custom-interval-set">Generate a custom interval set</h2>
<p>To build a custom set of intervals from a GTF file (here referred to as <code>genecode.gtf.gz</code>), you need to have <code>tabix</code> and <code>bgzip</code> installed. If you built a conda environment for the setup using the <code>setup_env.yml</code> file, <code>tabix</code> and <code>bgzip</code> will be present.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>zgrep<span class="w"> </span><span class="s2">&quot;protein_coding&quot;</span><span class="w"> </span>gencode.gtf.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-F<span class="s1">&#39;\t&#39;</span><span class="w"> </span><span class="s1">&#39;{ if ($3 == &quot;exon&quot;) print $1, $2, $3}&#39;</span><span class="w"> </span><span class="nv">OFS</span><span class="o">=</span><span class="s1">&#39;\t&#39;</span><span class="w"> </span>-<span class="w"> </span><span class="p">|</span><span class="w"> </span>bgzip<span class="w"> </span>-c<span class="w"> </span>&gt;<span class="w"> </span>calling_intervals.BED.gz<span class="w"> </span><span class="o">&amp;&amp;</span><span class="se">\</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>tabix<span class="w"> </span>-p<span class="w"> </span>bed<span class="w"> </span>calling_intervals.BED.gz
</code></pre></div>
<p>Then you can use the <code>calling_intervals.BED.gz</code> file as the input for the variant calling step, by putting it in the <code>config_main.yaml</code></p>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; 2025 <a href="https://github.com/ctglab"  target="_blank" rel="noopener">Computational and Translational Genomics Laboratory (CTGLab)</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.sections", "toc.integrate", "navigation.top", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>